Logout ely 2534

rsearls opened the pull request:
https://issues.redhat.com/browse/ELY-2534

supersede https://github.com/wildfly-security/wildfly-elytron/pull/2245

pedroigor committed:
OpenID Connect Logout support

fjuma committed:
[ELY-2534] Update copyright headers

fjuma committed:
[ELY-2534] Update the creation of the test HTTP server request for back-channel logout

fjuma committed:
[ELY-2534] Update back-channel logout handling so it doesn't rely on an active session

rsearls committed:
logout file with debug stmts

rsearls committed:
[ELY-2534] OIDC logout support

rsearls committed:
working code for rpinitiated logout. contains tmp debug stmts

fjuma committed:
[ELY-2534] Restore the changes made so that back-channel logout handling doesn't rely on an active session and ensure the session will get invalidated appropriately upon subsequent requests

fjuma committed:
[ELY-2534] Fix back-channel logout by moving the isSessionMarkedForInvalidation check to before the authenticator#authenticate call and update the check to not remove the session from the map to ensure that the user gets logged out from any other apps too

fjuma committed:
[ELY-2534] Use the client ID and sid from the ID token in the key for the sessionsMarkedForInvalidation map

pedro-hos commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,272 @@
+    private static final String POST_LOGOUT_REDIRECT_URI_PARAM = "post_logout_redirect_uri";
+    private static final String ID_TOKEN_HINT_PARAM = "id_token_hint";
+    private static final String LOGOUT_TOKEN_PARAM = "logout_token";
+    private static final String LOGOUT_TOKEN_TYPE = "Logout";
```
"Logout" with the first capital letter is the expected value?

rsearls replied:
Yes "Logout" is expected it looks to be a key for claim types.  similarly "Bearer" is capitalized as well.

I changed the log.debug ref to log.debugf only in the code I added.  I prefer to leave pre-existing code as is.  

Thanks for the code review and the suggestions.

fjuma replied:
Are "logout_token" and "Logout" specific to the Keycloak OpenID provider? I'm not seeing these in the OIDC logout specs. If these are Keycloak-specific, we need to be careful to make sure we are not introducing logic that won't work for providers that don't make use of these.

That reminds me, it would also be good to do some manual tests using your built WildFly server with other OpenID providers like Auth0 or Okta to make sure things work as expected with a non-Keycloak provider.

rsearls replied:
"logout_token" is defined in the Back-Channel spec, so it is not Elytron specific.

fjuma replied:
I think the "Logout" value for the "typ" claim looks to be specific to Keycloak. @pedroigor Can you confirm?

In the Back Channel spec, I see the following:

"Another way to prevent cross-JWT confusion is to use explicit typing, as described in Section 3.11 of [[RFC8725]](https://openid.net/specs/openid-connect-backchannel-1_0.html#RFC8725). One would explicitly type a Logout Token by including a typ (type) Header Parameter with a value of logout+jwt (which is registered in [Section 5.3.1](https://openid.net/specs/openid-connect-backchannel-1_0.html#MediaContents). Note that the application/ portion of the application/logout+jwt media type name is omitted when used as a typ Header Parameter value, as described in the typ definition in Section 4.1.9 [[JWS]](https://openid.net/specs/openid-connect-backchannel-1_0.html#JWS). Including an explicit type in issued Logout Tokens is a best practice. Note however, that requiring explicitly typed Logout Tokens will break most existing deployments, as existing OPs and RPs are already commonly using untyped Logout Tokens. However, requiring explicit typing would be a good idea for new deployment profiles where compatibility with existing deployments is not a consideration."

https://openid.net/specs/openid-connect-backchannel-1_0.html#CrossJWT

pedro-hos reviewed (commented):
LGTM overall! One small improvement could be adjusting the debug message to match the format of other log statements. For instance, instead of:  

```java
log.debug("evaluateRequest uri: " + request.getRequestURI().toString());
```

You might consider using `log.debugf` for consistency:  

```java
log.debugf("The evaluated request URI is [%s]", request.getRequestURI().toString());
```

This keeps the style aligned with existing logs, like:  

```java
log.debugf("Ignoring request for path [%s] from mechanism [%s]. No client configuration context found." ...);
```
That said, this is a minor improvement — everything else looks great to me!

pedro-hos reviewed (approved):
LGTM

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -63,19 +63,24 @@ protected boolean removeEldestEntry(Map.Entry<String, OidcClientConfiguration> e
     });
 
     boolean tryLogout(OidcHttpFacade facade) {
+        log.trace("tryLogout entered");
```
Just to check, I know some of these trace and debug statements were added for debugging. Do we still want to keep all of these?

rsearls replied:
I can remove them but I think they could be very helpful debugging customer issues with logout.

fjuma replied:
Keeping ones that make sense is fine. Might be good to revisit the messages and make sure they have sufficient details/are consistent.

rsearls replied:
done

fjuma commented on tests/base/src/test/java/org/wildfly/security/http/impl/AbstractBaseHttpTest.java:
```diff
@@ -358,6 +358,9 @@ public boolean resumeRequest() {
         public HttpScope getScope(Scope scope) {
+            if (requestURI != null && "/clientApp/logout/callback".equals(requestURI.getPath())){
+                return null;
+            }
```
Why is this needed here?

rsearls replied:
This is needed to allow assertUserNotAuthenticated() after user logout to function properly.

fjuma replied:
Just wondering if this is the right place to add this logic since getScope doesn't take the requestURI as a param.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/ElytronMessages.java:
```diff
@@ -278,5 +278,12 @@ interface ElytronMessages extends BasicLogger {
     @Message(id = 23070, value = "Authentication request format must be one of the following: oauth2, request, request_uri.")
     RuntimeException invalidAuthenticationRequestFormat();
+
+    @Message(id = 23071, value = "%s is not a valid value for %s")
```
Once the nonce changes are merged, we'll need to merge the latest changes from 2.x into this branch and these message IDs will need to be updated here.

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfiguration.java:
```diff
@@ -143,9 +143,9 @@ public enum RelativeUrlsUsed {
     protected String requestObjectSigningKeyStoreType;
     protected JWKEncPublicKeyLocator encryptionPublicKeyLocator;
-    private boolean logoutSessionRequired = true;
+    private String logoutSessionRequired;
```
Why was this changed to a String?

rsearls replied:
This is an optional setting.  We needed to be able to determine if the user specifically set the value or not.
If this was a boolean we would not be able to determine if the user set the value to true, false or not set at all.

fjuma replied:
From the analysis document, it looks like the default for this should be set to true. We should be using `boolean` here (we have other similar boolean attributes as well). (Note that it's also possible to use `Boolean` but I think that will be needed for the system property case, it shouldn't be needed here.)

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfigurationBuilder.java:
```diff
@@ -199,7 +199,10 @@ protected OidcClientConfiguration internalBuild(final OidcJsonConfiguration oidc
-        String tmpLogoutPath = System.getProperty(LOGOUT_PATH);
+        String tmpLogoutPath = oidcJsonConfiguration.getLogoutPath();
+        if (tmpLogoutPath == null) {
+            tmpLogoutPath = System.getProperty(LOGOUT_PATH);
```
The system property configuration shouldn't be part of this branch. In this 2.x branch, logout should be configured using OIDC JSON configuration only.

rsearls replied:
done

fjuma commented:
In https://github.com/wildfly-security/wildfly-elytron/pull/2249/commits/bbc858cc8044755b4e8b7339d522ae3bb014438d, it looks like there are changes to tool/LICENSE.txt, there shouldn't be any changes there.

fjuma commented on http/oidc/src/test/java/org/wildfly/security/http/oidc/LogoutConfigurationOptionsTest.java:
```diff
@@ -0,0 +1,109 @@
+    public void testLogoutPath() {
+
+        try {
+            System.setProperty(Oidc.LOGOUT_PATH, " ");
```
System property related tests should be for the maintenance branch. They shouldn't be present in this 2.x branch.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/AuthenticatedActionsHandler.java:
```diff
@@ -1,6 +1,6 @@
 /*
  * JBoss, Home of Professional Open Source.
- * Copyright 2021 Red Hat, Inc., and individual contributors
+ * Copyright 2024 Red Hat, Inc., and individual contributors
```
Just a minor comment, we don't need to update the year on existing files.

rsearls replied:
original date returned

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/ElytronMessages.java:
```diff
@@ -278,5 +278,11 @@ interface ElytronMessages extends BasicLogger {
     @Message(id = 23070, value = "Authentication request format must be one of the following: oauth2, request, request_uri.")
     RuntimeException invalidAuthenticationRequestFormat();
+
+    @Message(id = 23071, value = "Invalid logout output: %s is not a valid value for %s")
```
Now that the OIDC nonce PR was merged, it's a good time to add a commit that merges the latest changes from 2.x and then update the message IDs.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfiguration.java:
```diff
@@ -692,6 +704,22 @@ public String getTokenSignatureAlgorithm() {
+        this.sessionRequiredOnLogout = sessionRequiredOnLogout;
+    }
+
+    public int getLogoutSessionWaitingLimit() {
```
Just thinking it might be good to rename this to make it easier to understand what it's used for.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,270 @@
+            logoutUri = redirectUriBuilder.build().toString();
+            log.trace("redirectEndSessionEndpoint path: " + redirectUriBuilder.toString());
+        } catch (URISyntaxException e) {
+            throw new RuntimeException(e);
```
We should use ElytronMessages here.

rsearls replied:
Done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,270 @@
+        JwtClaims claims;
+
+        try {
+            claims = tokenValidator.verify(logoutToken);
```
As mentioned above, we need to check if this "Logout" token type is Keycloak-specific because if it is, this validation will fail for OpenID providers other an than Keycloak.

rsearls replied:
"logout_token" is defined in the Back-Channel spec, so it is not Elytron specific.

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,270 @@
+        }
+
+        if (!isSessionRequiredOnLogout(facade)) {
+            log.warn("Back-channel logout request received but can not infer sid from logout token to mark it for invalidation");
```
This should be in ElytronMessages.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/Oidc.java:
```diff
@@ -173,6 +173,10 @@ public class Oidc {
     public static final String PROVIDER_URL = "provider-url";
+    public static final String LOGOUT_PATH = "logout-path";
+    public static final String LOGOUT_CALLBACK_PATH = "logout-callback-path";
+    public static final String POST_LOGOUT_URI = "post-logout-uri";
```
I think the RP initiated logout spec refers to this as post_logout_redirect_uri. Any reason to not use `POST_LOGOUT_REDIRECT_URI=post-logout-redirect-uri` here?

rsearls replied:
This can be changed but it will required a revision to the proposal doc,
wildfly/elytron-oidc-client schema doc and corresponding code, and the
integration tests.   How would you like to proceed with this?

fjuma replied:
In general, it's good to use names that are consistent with the spec so it's intuitive how things defined in the spec relate to things in our configuration. I think the original name that was used in the proposal was `post-logout-redirect-uri`. Do you recall why it was changed to to `post-logout-uri`?

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfiguration.java:
```diff
@@ -143,6 +143,14 @@ public enum RelativeUrlsUsed {
+    private String logoutPath = "/logout";
+    private String logoutCallbackPath = "/logout/callback";
+
+    private int logoutSessionWaitingLimit = 100;
```
Would be good to define a constant in Oidc.java for this default value.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfiguration.java:
```diff
@@ -143,6 +143,14 @@ public enum RelativeUrlsUsed {
+    private String postLogoutUri;
+    private boolean sessionRequiredOnLogout = true;
+    private String logoutPath = "/logout";
+    private String logoutCallbackPath = "/logout/callback";
```
Would be good to define a constant in Oidc.java for the above two default values.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfiguration.java:
```diff
@@ -323,9 +331,13 @@ public String getTokenUrl() {
         return tokenUrl;
     }
 
-    public String getLogoutUrl() {
```
Is this rename needed?

rsearls replied:
The value being return is the path component of a URL not the URL itself.
For accuracy and clarity the name was changed.   In addition this variable
is directly related to the property name logout-path as defined in the 
proposal doc and the elytron-oidc-client schema.

fjuma replied:
Sorry, I was asking about the getLogoutUrl method being renamed to getEndSessionEndpointUrl.

fjuma replied:
I see now that getEndSessionEndpointUrl wasn't part of your commits, it was actually part of Pedro's initial changes. That's fine.

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfiguration.java:
```diff
@@ -779,4 +807,32 @@ public void setEncryptionPublicKeyLocator(JWKEncPublicKeyLocator publicKeySetExt
         return this.encryptionPublicKeyLocator;
     }
+
+    public void setPostLogoutUri(String postLogoutUri) {
```
Could we use use `setPostLogoutRedirectUri(String postLogoutRedirectUri)`?

rsearls replied:
Elytron's coding convention appears to be to define a get/set method name
that uses the same variable name.  If we change variable postLogoutUri to 
postLogoutRedirectUri as discussed in a previous comment then this method 
name should change as well.

fjuma replied:
Yes, that's right, just wanted to point these references out as a reminder to myself when reviewing. Sometimes it's easy to update a variable name but forget to update some of the methods that reference it. :)

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfiguration.java:
```diff
@@ -779,4 +807,32 @@ public void setEncryptionPublicKeyLocator(JWKEncPublicKeyLocator publicKeySetExt
+        this.postLogoutUri = postLogoutUri;
+    }
+
+    public String getPostLogoutUri() {
```
Similar comment here as the one above.

rsearls replied:
Change dependent on previous name decision as discussed above

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfigurationBuilder.java:
```diff
@@ -195,6 +198,38 @@ protected OidcClientConfiguration internalBuild(final OidcJsonConfiguration oidc
         oidcClientConfiguration.setTokenSignatureAlgorithm(oidcJsonConfiguration.getTokenSignatureAlgorithm());
 
+        String tmpLogoutPath = oidcJsonConfiguration.getLogoutPath();
+        log.debugf("sysProp LOGOUT_PATH: " + (tmpLogoutPath == null ? "NULL" : tmpLogoutPath));
```
This references a system property.

fjuma replied:
The special null handling isn't needed.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfigurationBuilder.java:
```diff
@@ -195,6 +198,38 @@ protected OidcClientConfiguration internalBuild(final OidcJsonConfiguration oidc
+        }
+
+        String tmpLogoutCallbackPath = oidcJsonConfiguration.getLogoutCallbackPath();
+        log.debugf("sysProp LOGOUT_CALLBACK_PATH: " + (tmpLogoutCallbackPath == null ? "NULL" : tmpLogoutCallbackPath));
```
This should be updated, currently references sysProp.

fjuma replied:
The special null handling isn't needed.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfigurationBuilder.java:
```diff
@@ -195,6 +198,38 @@ protected OidcClientConfiguration internalBuild(final OidcJsonConfiguration oidc
+        String tmpLogoutCallbackPath = oidcJsonConfiguration.getLogoutCallbackPath();
+        log.debugf("sysProp LOGOUT_CALLBACK_PATH: " + (tmpLogoutCallbackPath == null ? "NULL" : tmpLogoutCallbackPath));
+        if (tmpLogoutCallbackPath != null) {
+            if (tmpLogoutCallbackPath.startsWith("http")) {
```
I don't think this check is needed. In the spec, I see, "This URI SHOULD use the https scheme...it MAY use the http scheme...The URI MAY use an alternate scheme..."

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfigurationBuilder.java:
```diff
@@ -195,6 +198,38 @@ protected OidcClientConfiguration internalBuild(final OidcJsonConfiguration oidc
+        }
+
+        String tmpPostLogoutUri = oidcJsonConfiguration.getPostLogoutUri();
+        log.debugf("sysProp POST_LOGOUT_URI: " + (tmpPostLogoutUri == null ? "NULL" : tmpPostLogoutUri));
```
This should be updated, currently references sysProp.

fjuma replied:
The special null handling isn't needed.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfigurationBuilder.java:
```diff
@@ -236,4 +271,11 @@ public static OidcClientConfiguration build(OidcJsonConfiguration oidcJsonConfig
 
+    private boolean isValidPath(String path) {
+        String tmpPath = path.trim();
+        if (tmpPath.length() > 1 && tmpPath.startsWith("/")) {
```
Just a minor comment, this could be simplified to `return tmpPath.length() > 1 && tmpPath.startsWith("/");`

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcJsonConfiguration.java:
```diff
@@ -99,8 +103,9 @@
         PROXY_URL, TURN_OFF_CHANGE_SESSION_ID_ON_LOGIN, TOKEN_MINIMUM_TIME_TO_LIVE,
-        MIN_TIME_BETWEEN_JWKS_REQUESTS, PUBLIC_KEY_CACHE_TTL,
-        IGNORE_OAUTH_QUERY_PARAMETER, VERIFY_TOKEN_AUDIENCE, TOKEN_SIGNATURE_ALGORITHM, SCOPE,
+        MIN_TIME_BETWEEN_JWKS_REQUESTS, POST_LOGOUT_URI, PUBLIC_KEY_CACHE_TTL,
```
Just a minor comment, I don't think these are in alphabetical order so we could just add the new param at the end of the list to make it easier to see the changes.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcJsonConfiguration.java:
```diff
@@ -152,13 +157,21 @@ public class OidcJsonConfiguration {
     @JsonProperty(IGNORE_OAUTH_QUERY_PARAMETER)
     protected boolean ignoreOAuthQueryParameter = false;
+    @JsonProperty(LOGOUT_PATH)
+    protected String logoutPath;
```
Could set the default value here and below.

rsearls replied:
Done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcJsonConfiguration.java:
```diff
@@ -152,13 +157,21 @@ public class OidcJsonConfiguration {
+    protected String logoutCallbackPath;
+    @JsonProperty(LOGOUT_SESSION_REQUIRED)
+    protected boolean logoutSessionRequired = true;
     @JsonProperty(VERIFY_TOKEN_AUDIENCE)
```
I think we're missing something for what's currently called `logoutSessionWaitingLimit` in this file.

rsearls replied:
This is an Elytron specific setting.  There is no mention of this 
and any of the specs.   Still add it to OidcJsonConfiguration and
other structures as needed?

fjuma replied:
Since it's part of the OidcClientConfiguration, it should be present here. Maybe check with @pedroigor to see if this is something that should be configurable or not. I believe this was part of the original analysis document so I thought it was meant to be configurable.

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/TokenValidator.java:
```diff
@@ -110,11 +112,17 @@ public VerifiedTokens parseAndVerifyToken(final String idToken, final String acc
+        return new AccessToken(verify(bearerToken));
+    }
+
+    public JwtClaims verify(String bearerToken) throws OidcException {
```
s/bearerToken/token

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/TokenValidator.java:
```diff
@@ -123,15 +131,15 @@ public AccessToken parseAndVerifyToken(final String bearerToken) throws OidcExce
-            JwtClaims jwtClaims = jwtContext.getJwtClaims();
+            jwtClaims = jwtContext.getJwtClaims();
             if (jwtClaims == null) {
                 throw log.invalidBearerTokenClaims();
```
Because we are changing the method to take a logout token, not necessarily only bearer tokens as before, we might need to adjust the logging.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/TokenValidator.java:
```diff
@@ -213,11 +224,24 @@ public TokenValidator build() throws IllegalArgumentException {
+            return this;
+        }
+
+        public Builder setTokenType(String tokenType) {
```
Just wondering if this method is needed?

rsearls replied:
Yes, it's used in LogoutHandler. The default value, Bearer, for tokenType must be overridden.

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/TokenValidator.java:
```diff
@@ -164,13 +172,16 @@ public static Builder builder(OidcClientConfiguration clientConfiguration) {
 
     public static class Builder {
+
+        public String tokenType = "Bearer";
```
I believe "Bearer" is a Keycloak specific value. Just wondering if we need to set tokenType at all.

rsearls replied:
Yes we need it.  It's used by jwtConsumerBuilder for the type of ErrorCodeValidator

fjuma replied:
Just wondering why it's needed in the builder. Will need to take a closer look.

rsearls commented:
Implemented code changes
- replace post-logout-uri with post-logout-redirect-uri
- added property back-channel-logout-session-invalidation-limit
- changed logoutSessionWaitingLimit to backChannelLogoutSessionInvalidationLimit
- removed sessionRequiredOnLogout using logoutSessionRequired instead

"typ" has not been addressed.

rsearls commented:
Replace "Logout" with "logout+jwt".

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -151,7 +151,7 @@ private void handleBackChannelLogoutRequest(OidcHttpFacade facade) {
         TokenValidator tokenValidator = TokenValidator.builder(facade.getOidcClientConfiguration())
                 .setSkipExpirationValidator()
-                .setTokenType(LOGOUT_TOKEN_TYPE)
+                .setTokenType(LOGOUT_JWT_TOKEN_TYPE)
```
With this change, won't logout fail when using Keycloak? We need a solution that will work for both Keycloak and other OpenID providers. I see the tests passing though. Have you checked to see how they are passing if the Keycloak logout token contains typ with value Logout?

rsearls replied:
I've tested this with the wfly integration tests and it does not fail.  Keycloak must be supporting both.

fjuma replied:
Please debug to see what Keycloak's logout token looks like and what its claims are. It seems very strange that validation would pass before with checking the typ claim is set to "Logout" and now validation still passes with the typ claim set to "logout+jwt" instead. Is the validator actually getting invoked at all?

We also need to be careful because the spec says the typ value should be "logout+jwt", it doesn't say it MUST be set to this value.

rsearls replied:
There is no way to determine which OpenID Provider is being used and thus
which claim 'typ' to use.  I have added code to use keycloak's 'typ' first.  
If that throws an exception the typ 'logout+jwt' is tried.

fjuma replied:
Were you able to determine why the tests happened to be passing with Keycloak when attempting to validate that typ was set to 'logout+jwt'?

rsearls replied:
I believe that code was not being called because isLogoutCallbackPath was not performing the proper check.

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/Oidc.java:
```diff
@@ -180,7 +180,14 @@ public class Oidc {
+    public static final String LOGOUT_SESSION_REQUIRED = "logout-session-required";
+    public static final String logoutPath = "/logout";
+    public static final String logoutCallbackPath = "/logout/callback";
+    public static final int backChannelLogoutSessionInvalidationLimit = 16;
```
We should use upper case for the above three variable names since these are constants. We can also prefix with DEFAULT, e.g., DEFAULT_LOGOUT_PATH="/logout"

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/Oidc.java:
```diff
@@ -180,7 +180,14 @@ public class Oidc {
+    public static final String logoutPath = "/logout";
+    public static final String logoutCallbackPath = "/logout/callback";
+    public static final int backChannelLogoutSessionInvalidationLimit = 16;
+    public static final String BACK_CHANNEL_LOGOUT_SESSION_INVALIDATION_LIMIT = "back-channel-logout-session-invalidation-limit";
```
DEFAULT_BACK_CHANNEL_LOGOUT_SESSION_INVALIDATION_LIMIT

fjuma replied:
For all of the above constants, I don't think we need to make them public, just `static final` should be good.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfigurationBuilder.java:
```diff
@@ -195,6 +196,31 @@ protected OidcClientConfiguration internalBuild(final OidcJsonConfiguration oidc
         oidcClientConfiguration.setTokenSignatureAlgorithm(oidcJsonConfiguration.getTokenSignatureAlgorithm());
 
+        String tmpLogoutPath = oidcJsonConfiguration.getLogoutPath();
+        log.debugf("property LOGOUT_PATH: " + tmpLogoutPath);
```
The references to "property" in the debug statements should be removed.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfigurationBuilder.java:
```diff
@@ -195,6 +196,31 @@ protected OidcClientConfiguration internalBuild(final OidcJsonConfiguration oidc
 
+        String tmpLogoutPath = oidcJsonConfiguration.getLogoutPath();
+        log.debugf("property LOGOUT_PATH: " + tmpLogoutPath);
+        if (tmpLogoutPath != null) {
```
Just noticed that in the rest of this class, it's common to set the default value in the else statement, e.g.,

```
if (oidcJsonConfiguration.getLogoutPath() != null) {
...
} else {
    oidcClientConfiguration.setLogoutPath(DEFAULT_LOGOUT_PATH)
}
```

fjuma replied:
The same could be done for each of the options below as well.

rsearls replied:
done

fjuma commented on tool/LICENSE.txt:
```diff
@@ -1 +0,0 @@
-src/main/resources/META-INF/LICENSE.txt
```
This shouldn't be removed.

rsearls replied:
done

skyllarr replied:
@rsearls This file should not be altered? It contains a single line originally: https://github.com/wildfly-security/wildfly-elytron/blob/2.x/tool/LICENSE.txt

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/ElytronMessages.java:
```diff
@@ -287,5 +287,17 @@ interface ElytronMessages extends BasicLogger {
     @Message(id = 23073, value = "Nonce cookie does not exist")
     String nonceCookieDoesNotExist();
+
+    @Message(id = 23074, value = "Invalid logout output: %s is not a valid value for %s")
```
s/output/path

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/ElytronMessages.java:
```diff
@@ -287,5 +287,17 @@ interface ElytronMessages extends BasicLogger {
     String nonceCookieDoesNotExist();
+
+    @Message(id = 23074, value = "Invalid logout output: %s is not a valid value for %s")
+    RuntimeException invalidLogoutPath(String pathValue, String pathName);
```
I think we should use IllegalArgumentException here.

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/ElytronMessages.java:
```diff
@@ -287,5 +287,17 @@ interface ElytronMessages extends BasicLogger {
+    RuntimeException invalidLogoutPath(String pathValue, String pathName);
+
+    @Message(id = 23075, value = "Invalid %s: %s the value must be an absolute URI")
+    RuntimeException invalidLogoutCallbackPath(String callbackPathTitle, String callbackPathValue);
```
IllegalArgumentException

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/ElytronMessages.java:
```diff
@@ -287,5 +287,17 @@ interface ElytronMessages extends BasicLogger {
+    @Message(id = 23074, value = "Invalid logout output: %s is not a valid value for %s")
+    RuntimeException invalidLogoutPath(String pathValue, String pathName);
+
+    @Message(id = 23075, value = "Invalid %s: %s the value must be an absolute URI")
```
s/the value/the logout callback path value

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,271 @@
+            }
+
+            logoutUri = redirectUriBuilder.build().toString();
+            log.trace("redirectEndSessionEndpoint path: " + redirectUriBuilder.toString());
```
Can logoutUri be used here instead of redirectUriBuilder.toString()?

rsearls replied:
done

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,271 @@
+            log.trace("redirectEndSessionEndpoint path: " + redirectUriBuilder.toString());
+        } catch (URISyntaxException e) {
+            throw log.invalidOpenIdProviderLogoutEndpoint(
+                    clientConfiguration.getEndSessionEndpointUrl(), e.getMessage());
```
If we get here, either there's an issue with the endSessionEndpointUrl itself or with the ID token hint value or the post_logout_redirect_uri value, right? Maybe instead of invalidOpenIDProviderLogoutEndpoint, we could use something like unableToCreateEndSessionEndpointRequest and update the message itself accordingly.

rsearls replied:
done

fjuma commented on http/oidc/src/test/java/org/wildfly/security/http/oidc/AbstractLogoutTest.java:
```diff
@@ -0,0 +1,241 @@
+                CLIENT_APP, false);
+
+        realm.setAccessTokenLifespan(100);
+        realm.setSsoSessionMaxLifespan(100);
```
I think these were set to high values to ensure that logout is actually succeeding and it's not just that the token has expired or something like that right? It would be good to add a comment about that here so we remember this later.

rsearls replied:
done

fjuma commented on http/oidc/src/test/java/org/wildfly/security/http/oidc/AbstractLogoutTest.java:
```diff
@@ -0,0 +1,241 @@
+    protected void assertUserNotAuthenticated() {
+        URI requestURI = getCurrentRequest().getRequestURI();
+        if (requestURI != null && "/clientApp/logout/callback".equals(requestURI.getPath())){
+            assertNull(null);
```
What are we trying to assert here?

rsearls replied:
fixed

rsearls commented:
Revisions pushed to PR

fjuma commented on http/oidc/src/test/java/org/wildfly/security/http/oidc/BackChannelLogoutTest.java:
```diff
@@ -0,0 +1,80 @@
+        String redirectUri = redirectUris.get(0);
+
+        OidcClientConfiguration config = new OidcClientConfiguration();
+        config.setLogoutCallbackPath(Oidc.DEFAULT_LOGOUT_CALLBACK_PATH);
```
Do we have a test for the case where the logout callback path is an absolute URL?

rsearls replied:
Added

fjuma commented on http/oidc/src/test/java/org/wildfly/security/http/oidc/FrontChannelLogoutTest.java:
```diff
@@ -0,0 +1,129 @@
+        List<String> redirectUris = client.getRedirectUris();
+        String redirectUri = redirectUris.get(0);
+        OidcClientConfiguration config = new OidcClientConfiguration();
+        config.setLogoutCallbackPath(Oidc.DEFAULT_LOGOUT_CALLBACK_PATH);
```
Same here, do we have a test for the case where the callback path is an absolute URL?

rsearls replied:
added

rsearls commented:
added tests pushed

fjuma commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,303 @@
+    private boolean isLogoutCallbackPath(OidcHttpFacade facade) {
+        String uriStr = facade.getRequest().getURI();
+        // logoutCallbackPath can be either an URL path component or an absolute path.
+        // Only the path components are to be compared.
```
Just wondering, if the logoutCallbackPath is an absolute URL, shouldn't we check that it is the same as uriStr?

Previously, only the path components were compared because we were using just the path.

rsearls replied:
changed

fjuma commented on http/oidc/src/test/java/org/wildfly/security/http/oidc/AbstractLogoutTest.java:
```diff
@@ -0,0 +1,240 @@
+        config.setRealm(TEST_REALM);
+        config.setResource(CLIENT_ID);
+        config.setPublicClient(false);
+        config.setAuthServerUrl(KEYCLOAK_CONTAINER.getAuthServerUrl());
```
Would be good to use setProviderUrl and setClientId instead of setRealm, setResource, and setAuthServerUrl

rsearls replied:
changed

rsearls commented:
PR updated

beth-soptim commented:
Hi, do you have any updates for this feature? Last year or so it was once planned for Wildfly 37.

rsearls commented:
There is no specific version for release yet but it is in the planning queue.

skyllarr commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/ElytronMessages.java:
```diff
@@ -211,8 +211,8 @@ interface ElytronMessages extends BasicLogger {
 
-    @Message(id = 23050, value = "Invalid bearer token claims")
-    OidcException invalidBearerTokenClaims();
+    @Message(id = 23050, value = "Invalid token claims")
```
@rsearls Can we add another message instead of changing the existing one? For some backwards compatibility in logging

rsearls replied:
switching the msg order addresses the issue.

skyllarr replied:
@rsearls Can you pls clarify how this was addressed? it is just a minor, but I think that the message with ID 23050 already existed prior to this feature, so instead of editing it, we could add another message with new unique ID

skyllarr replied:
@rsearls  actually looking more closely, the only place where this is used is in TokenValidator.  This validates both the bearer token and ID token so it might have been renamed on purpose to address both token types.  Or do we know why was this renamed? Was it renamed by somebody esle in commits?

rsearls replied:
I reviewed the code history and saw this before yesterdays change.  
        214       @Message(id = 23050, value = "Invalid bearer token claims")
        215       OidcException invalidBearerTokenClaims();
            214   @Message(id = 23050, value = "Invalid token claims")
            215   OidcException invalidTokenClaims();
The first usage, "invalidBearerTokenClaims" was replaced by "invalidTokenClaims"  on 3/21/25.  
This was done because the message itself was changed from a reference specifically 
to "bearer token" to a more generic message; the message method naming pattern 
was being followed.  A method call to "invalidBearerTokenClaims" no longer exists
in the code base.  Most likely the message was made more generic when an 
addition of nonce handling was added.  Most likely the old message was returned 
when a merge of elytron 2.x was performed last week to bring this branch up to date.

skyllarr replied:
@rsearls Hm when I look at the https://github.com/wildfly-security/wildfly-elytron/blame/2.x/http/oidc/src/main/java/org/wildfly/security/http/oidc/ElytronMessages.java#L215C32-L215C43 , it says that the 2.x branch has the invalidBearerTokenClaims since the bearer token support was added 3 years ago. Do you maybe have a different view locally?

rsearls replied:
The name was generalize when BackChannelLogout logoutToken
validation was added.
I think when I did the rebase of 2.x to my branch last week is when the competing messages appeared.

skyllarr commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/ElytronMessages.java:
```diff
@@ -287,5 +287,17 @@ interface ElytronMessages extends BasicLogger {
     String nonceCookieDoesNotExist();
+
+    @Message(id = 23074, value = "Invalid logout path: %s is not a valid value for %s")
+    IllegalArgumentException invalidLogoutPath(String pathValue, String pathName);
```
just a minor, I think this message and below message can accept a single parameter, as the logout path will be an attribute and always have a value of "logout-path" ?

rsearls replied:
+    @Message(id = 23074, value = "Invalid logout path: %s is not a valid value for %s")
+    IllegalArgumentException invalidLogoutPath(String pathValue, String pathName);

For clarity for the user I think the current message should be maintained
or the message changed to "%s is not a valid value for attribute logout-path".
What is your preference?

I have removed "invalidLogoutCallbackPath" as it is not being used anywhere

skyllarr replied:
@rsearls  My thought was changing it to be `"%s is not a valid value for attribute logout-path".`  Or it could also be simplified to be more generic and used in more situations:

```
@Message(id = 23074, value = "Invalid attribute: %s is not a valid value for %s")
IllegalArgumentException invalidAttribute(String pathValue, String pathName);
```

But thinking more about it, since this is a preview feature, and there is a tiny possibility of the "logout-path" changing in the future, let's keep it as it is now.

Btw I can still see the "invalidLogoutCallbackPath" in these changes, so it was not removed.

rsearls replied:
I was waiting to hear your preference for the message in "invalidAttribute" before checking in the change where I removed the  "invalidLogoutCallbackPath"

skyllarr commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/ElytronMessages.java:
```diff
@@ -287,5 +287,17 @@ interface ElytronMessages extends BasicLogger {
+    RuntimeException unableToCreateEndSessionEndpointRequest(String url, String msg);
+
+    @Message(id = 23077, value = "Back-channel logout request received but can not infer sid from logout token to mark it for invalidation")
+    String sidCanNotBeInferFromLogoutToken();
```
s/sidCanNotBeInferFromLogoutToken/sidCanNotBeInferredFromLogoutToken

rsearls replied:
fixed typo

skyllarr commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,313 @@
+        if (idToken == null) {
+            return false;
+        }
+        return sessionsMarkedForInvalidation.remove(idToken.getNonce()) != null;
```
@rsearls We are removing the key "idToken.getNonce()" from the map. But we are putting "getSessionKey(facade, sessionId)" key there on line 209. 

This doesn't have to be the same key? Shouldn't we remove the getSessionKey(facade, sessionId) value here instead?

rsearls replied:
I think this is to high risk a change at this time.  I do not understand the
details of how IDToken or sessionId is used in these scenarios.  I think it
would be safest to create a jira for this to be addressed in the future.

skyllarr replied:
@rsearls I checked and we are the ones creating the NONCE value, send it with the initial login request to OpenID provider, and they send it back so we can validate it is the same value that we have sent. We are creating it in the OidcRequestAuthenticator [here](https://github.com/wildfly-security/wildfly-elytron/blob/2.x/http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcRequestAuthenticator.java#L643),  and then we compare the received nonce in TokenValidator [here](https://github.com/wildfly-security/wildfly-elytron/blob/2.x/http/oidc/src/main/java/org/wildfly/security/http/oidc/TokenValidator.java#L316). It seems you have added this code in the past so maybe you know whether my understanding is correct?

If so, how we use the map with different key being put there and different key removed seems to me like a bug, because the value of nonce will be random, but the  value of getSessionKey(facade, sessionId) will be something like clientId-sessionId.

rsearls replied:
Change implemented

rsearls replied:
Change implemented

skyllarr commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,313 @@
+    /**
+     * A bounded map to store sessions marked for invalidation after receiving logout requests through the back-channel
+     */
+    private Map<String, OidcClientConfiguration> sessionsMarkedForInvalidation = synchronizedMap(new LinkedHashMap<String, OidcClientConfiguration>(16, 0.75f, true) {
```
@rsearls Not sure if this was discussed, and it can possibly be an enhancement later. But this invalidation map will work only on the same JVM. Was there any discussion about whether it is makes sense to make it work in distributed environments?

rsearls replied:
There was never any discussion about this.  I recommend we create a jira to
evaluate and address this in the future.

skyllarr reviewed (commented):
@rsearls I haven't finish the review yet, just addint some in proress comments for now

rsearls commented:
In branch 2.x tool/LICENSE.txt is a link to a file.  I have addressed this.

skyllarr commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcClientConfiguration.java:
```diff
@@ -338,9 +345,13 @@ public String getTokenUrl() {
     }
 
-    public String getLogoutUrl() {
+    public String getEndSessionEndpointUrl() {
```
Just a note for other reviewers, http/oidc is not part of the public API so these renames are fine

skyllarr commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/TokenValidator.java:
```diff
@@ -115,33 +117,41 @@ public VerifiedTokens parseAndVerifyToken(final String idToken, final String acc
-            return new AccessToken(jwtClaims);
         } catch (InvalidJwtException e) {
-            log.tracef("Problem parsing bearer token: " + bearerToken, e);
+            log.tracef("Problem parsing bearer token: " + token, e);
```
@rsearls Just wondering why the message was changed from **invalidBearerTokenClaims** to  **invalidTokenClaims**  and if the log here should also have "**bearer**" removed and be: log.tracef("Problem parsing token: " + token, e);

skyllarr replied:
this relates to our other discussion above about whether we are changing this message in this PR or it was changed elsewhere already.

rsearls replied:
TokenVaidator.verify was generalized when BackChannelLogout logoutToken
validation was added.
    LogoutHandler.handleBackChannelLogoutRequest
    155  String logoutToken = facade.getRequest().getFirstParam(LOGOUT_TOKEN_PARAM);
            from the request object a logoutToken is looked for.  If present
            it contains jwtClaim parseable information related to logout.
    169  claims = tokenValidator.verify(logoutToken);
        
Token validation was no longer just for Bearer tokens.
    BearerTokenRequestAuthenticator.verifyToken
    99  token = tokenValidator.parseAndVerifyToken(tokenString);
        Method parseAndVerifyToken calls method verify
            TokenValidator.parseAndVerifyToken
            125 return new AccessToken(verify(token));

There is no reason "bearer" can't be remove from the log msg at
line 151.

Is that your preference?

skyllarr commented:
@rsearls Later today I plan to try to manually test all these logout mechanisms with keycloak locally. Just wondering if this is something you tried as well already?

skyllarr commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,313 @@
+        try {
+            URL url = new URL(tmpLogoutCallbackPath);
+            if (uriStr.equals(url.toString())
+                || uriStr.endsWith(tmpLogoutCallbackPath)) {
```
The  method getRelativePath() returns path that is relative to the context path of the application. So then IMO this should work: facade.getRequest().getRelativePath().equals(tmpLogoutCallbackPath)

rsearls replied:
I don't think that works in this usage.   Callback path is text that can be 
appended to a URL.  It is text that is registered with Keycloak as the post 
logout path and keycloak sends the url back on logout.

rsearls replied:
Change implemented

skyllarr commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,313 @@
+        if (logoutPath == null) {
+            return false;
+        }
+        return path.endsWith(logoutPath);
```
IMO here we can also use equals instead of endsWith

rsearls replied:
Yes this is for the same reason as above.

rsearls replied:
Changing this to equals caused tests to fail.

skyllarr commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,313 @@
+        facade.getTokenStore().logout(false);
+    }
+
+    private String getRedirectUri(OidcHttpFacade facade) {
```
Just a detail, this method is unused

rsearls replied:
I've removed it.

skyllarr commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,313 @@
+            IDToken idToken = context.getIDToken();
+            String issuer = request.getQueryParamValue(ISS);
+
+            if (idToken == null || !sessionId.equals(idToken.getNonce()) || !idToken.getIssuer().equals(issuer)) {
```
this is another instance of comparing `sessionId` (from `sid` query parameter) to `idToken.getNonce()` . Not sure  why since these are different concepts it seems to me. I wonder if Keycloak configures these values to match, but other providers might not match them. we will need to investigate this

rsearls replied:
agreed

skyllarr commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,313 @@
+    private void handleBackChannelLogoutRequest(OidcHttpFacade facade) {
+
+        OidcClientConfiguration clientConfiguration = facade.getOidcClientConfiguration();
+        String logoutToken = facade.getRequest().getFirstParam(LOGOUT_TOKEN_PARAM);
```
@rsearls Should we add a null check for logout token parameter here?

rsearls replied:
We could do a null check however an InvalidJwtException exception will be thrown by the TokenValidation code.  There is logging in place for that.

Do you want a null check to be added to the code any way?

skyllarr commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,313 @@
+
+        try {
+            URIBuilder redirectUriBuilder = new URIBuilder(clientConfiguration.getEndSessionEndpointUrl())
+                    .addParameter(ID_TOKEN_HINT_PARAM, securityContext.getIDTokenString());
```
maybe let's add a null check for securityContext.getIDTokenString() so that we do not add a parameter to the URI with null value

rsearls replied:
I've added the check.

rsearls commented:
I have not manually tested logout recently.  It was do in the past.

OndrejKotek commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/TokenValidator.java:
```diff
@@ -70,10 +70,12 @@ public Boolean run() {
     private static final int HEADER_INDEX = 0;
     private JwtConsumerBuilder jwtConsumerBuilder;
     private OidcClientConfiguration clientConfiguration;
+    private String tokenType;
```
What is/will be the field used for? I cannot find any read of its value.

rsearls replied:
tokenType is, "A JWT Claim typ (token type) is a header field in a 
JSON Web Token (JWT) that identifies the token as a JWT".
For a Bearer jwt claim it is Bearer"
For a logout jwt claim it is "logout+jwt"; this is defined in the
specs.

Line 180 of LogoutHandler sets this value for logout.

OndrejKotek replied:
Yes, it's set through TokenValidator.Builder. However the value seems not used in TokenValidator or anywhere else.

rsearls replied:
Addressed

OndrejKotek commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcAuthenticationMechanism.java:
```diff
@@ -58,6 +60,7 @@ public String getMechanismName() {
 
     @Override
     public void evaluateRequest(HttpServerRequest request) throws HttpAuthenticationException {
+        log.debugf("evaluateRequest uri: " + request.getRequestURI().toString());
```
```suggestion
        log.debugf("Evaluating request URI: [%s]", request.getRequestURI());
```

rsearls replied:
changed

OndrejKotek replied:
There are more of those. I'm not sure which are meant to stay.

There are also several `debugf` where `debug` would be sufficient instead.

rsearls replied:
All other such log msg are pre-existing.  I don't think they should be changed.

OndrejKotek commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,306 @@
+        OidcClientConfiguration clientConfiguration = facade.getOidcClientConfiguration();
+        String logoutToken = facade.getRequest().getFirstParam(LOGOUT_TOKEN_PARAM);
+        TokenValidator.Builder tokenBuilder = TokenValidator.builder(clientConfiguration)
+                .setSkipExpirationValidator();
```
Do I understand it correctly that we skip the Logout Token expiration validation? Is there another way to prevent captured Logout Tokens from being replayable we rely on?

rsearls replied:
I don't know.  Perhaps @pedroigor or @Skyllarr could comment on this.

OndrejKotek commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,306 @@
+        String sessionId = claims.getClaimValueAsString(SID);
+
+        if (sessionId == null) {
+            facade.getResponse().setStatus(HttpStatus.SC_BAD_REQUEST);
```
Does this mean that we do not support logout for all sessions (at once) at the RP for the End-User identified by the iss and sub Claims? If yes, it could be useful to add the fact to the analysis document.

rsearls replied:
Perhaps @pedroigor would like to comment on this.

OndrejKotek commented on http/oidc/src/main/java/org/wildfly/security/http/oidc/LogoutHandler.java:
```diff
@@ -0,0 +1,306 @@
+
+        OidcClientConfiguration clientConfiguration = facade.getOidcClientConfiguration();
+        String logoutToken = facade.getRequest().getFirstParam(LOGOUT_TOKEN_PARAM);
+        TokenValidator.Builder tokenBuilder = TokenValidator.builder(clientConfiguration)
```
I cannot find `events` Claim content validation, and not-presence of `nonce` Claim validation required by the specification.

rsearls replied:
There are various types of claims validators in TokenValidator.  There are several inner classes there are run for validation.

OndrejKotek replied:
Agree. Nonetheless it seems the validation required by the specification is missing.

rsearls replied:
Resolved

rsearls committed:
add check for tokenType

rsearls commented:
In preparation for a sprint.  Closing this PR.  Rebase branch to current 2.x and create a new PR
